src/engine/TransitivePathBase.cpp

```diff
@@ -45,36 +45,48 @@ TransitivePathBase::TransitivePathBase(
   }
   if (minDist_ == 0 && lhs_.isUnboundVariable() && rhs_.isUnboundVariable()) {
     emptyPathBound_ = true;
-    auto makeInternalVariable = [](std::string_view string) {
-      return Variable{
-          absl::StrCat("?internal_property_path_variable_", string)};
-    };
-    // Dummy variables to get a full scan of the index.
-    auto x = makeInternalVariable("x");
-    auto y = makeInternalVariable("y");
-    auto z = makeInternalVariable("z");
-    auto allValues = ad_utility::makeExecutionTree<Union>(
-        qec,
-        ad_utility::makeExecutionTree<IndexScan>(
-            qec, Permutation::Enum::SPO,
-            SparqlTriple{TripleComponent{x}, PropertyPath::fromVariable(y),
-                         TripleComponent{z}},
-            activeGraphs),
-        ad_utility::makeExecutionTree<IndexScan>(
-            qec, Permutation::Enum::OPS,
-            SparqlTriple{TripleComponent{z}, PropertyPath::fromVariable(y),
-                         TripleComponent{x}},
-            activeGraphs));
-    auto uniqueValues = ad_utility::makeExecutionTree<Distinct>(
-        qec, QueryExecutionTree::createSortedTree(std::move(allValues), {0}),
-        std::vector<ColumnIndex>{0});
-    lhs_.treeAndCol_.emplace(uniqueValues, 0);
+    lhs_.treeAndCol_.emplace(makeEmptyPathSide(qec, std::move(activeGraphs)),
+                             0);
   }
 
   lhs_.outputCol_ = 0;
   rhs_.outputCol_ = 1;
 }
 
+// _____________________________________________________________________________
+std::shared_ptr<QueryExecutionTree> TransitivePathBase::makeEmptyPathSide(
+    QueryExecutionContext* qec, Graphs activeGraphs) {
+  auto makeInternalVariable = [](std::string_view string) {
+    return Variable{absl::StrCat("?internal_property_path_variable_", string)};
+  };
+  // Dummy variables to get a full scan of the index.
+  auto x = makeInternalVariable("x");
+  auto y = makeInternalVariable("y");
+  auto z = makeInternalVariable("z");
+  // We don't need to materialize the extra variables y and z in the union.
+  auto selectXVariable =
+      [&x](std::shared_ptr<QueryExecutionTree> executionTree) {
+        executionTree->getRootOperation()->setSelectedVariablesForSubquery({x});
+        return executionTree;
+      };
+  auto allValues = ad_utility::makeExecutionTree<Union>(
+      qec,
+      selectXVariable(ad_utility::makeExecutionTree<IndexScan>(
```

Comment:
Add at least a TODO that we can also propagate this selection to the level of the index scans.